### PR TITLE
[TS migration] Migrate IOURequestStartPage, IOURequestRedirectToStartPage, IOURequestStepDistance and IOURequestStepAmount pages to TypeScript

### PR DESCRIPTION
**Pull request by kosmydel**
### Fixed Issues

$ https://github.com/Expensify/App/issues/38914
PROPOSAL:

### Tests

1. Create a new money request:
   1. Use FAB -> Request money
   2. Create new requests using Manual and Distance tabs.
2. Create a new money request using composer.
   1. Click "+" next to the composer.
   2. Select "Request money".
   3. Type some value -> Next -> Request.
   4. Click on the freshly created request. 
   5. Click on the amount to edit it.

### Offline tests

N/A

### QA Steps

- [ ] Verify that no errors appear in the JS console
- [ ] Do the tests from the "Tests" section

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/104823336/5c972de3-9597-4824-95e6-3489ede7f659

</details>

<details>
<summary>Android: mWeb Chrome & iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/104823336/4699660e-9c65-4f96-8f73-64a5ac38daeb

</details>

<details>
<summary>iOS: Native</summary>

Uploading ios.mp4…

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/104823336/c6d283b6-3780-4f68-a907-82dc56d34e26

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/104823336/431a54f7-2a54-49c2-9e8c-747fc3001dc3

</details>

### CODE REVIEW

File: src/pages/iou/request/step/IOURequestStepAmount.tsx [hunk 1]
```diff
@@ -56,22 +46,22 @@ function IOURequestStepAmount({
...
     const iouRequestType = getRequestType(transaction);
     const isEditing = action === CONST.IOU.ACTION.EDIT;
     const isSplitBill = iouType === CONST.IOU.TYPE.SPLIT;
     const isEditingSplitBill = isEditing && isSplitBill;
-    const {amount: transactionAmount} = ReportUtils.getTransactionDetails(isEditingSplitBill && !lodashIsEmpty(splitDraftTransaction) ? splitDraftTransaction : transaction);
-    const {currency} = ReportUtils.getTransactionDetails(isEditing ? draftTransaction : transaction);
+    const {amount: transactionAmount} = ReportUtils.getTransactionDetails(isEditingSplitBill && !lodashIsEmpty(splitDraftTransaction) ? splitDraftTransaction : transaction) ?? {amount: 0};
+    const {currency} = ReportUtils.getTransactionDetails(isEditing ? draftTransaction : transaction) ?? {currency: CONST.CURRENCY.USD};
 
     useFocusEffect(
```
Comment (kosmydel) [inline src/pages/iou/request/step/IOURequestStepAmount.tsx:62]: I think I could use `useAutoFocusInput` here if needed.

File: src/libs/Navigation/types.ts [hunk 2]
```diff
@@ -432,6 +426,29 @@ type MoneyRequestNavigatorParamList = {
...
     };
+    [SCREENS.MONEY_REQUEST.CREATE]: {
+        iouType: ValueOf<typeof CONST.IOU.TYPE>;
+        reportID: string;
+        transactionID: string;
+
+        // These are not used in the screen, but are needed for the navigation
+        // for IOURequestStepDistance and IOURequestStepAmount components
+        backTo: never;
+        action: never;
```
Comment (blazejkustra) [inline src/libs/Navigation/types.ts:437]: I removed it locally and typecheck works fine, how about removing it?

File: src/libs/actions/IOU.ts [hunk 3]
```diff
@@ -2172,15 +2172,24 @@ function updateMoneyRequestTaxRate(
...
-    policyTagList: OnyxEntry<OnyxTypes.PolicyTagList>,
-    policyCategories: OnyxEntry<OnyxTypes.PolicyCategories>,
-) {
+function updateMoneyRequestDistance({
+    transactionID,
+    transactionThreadReportID,
+    waypoints,
+    policy = {} as OnyxTypes.Policy,
+    policyTagList = {},
+    policyCategories = {},
```
Comment (blazejkustra) [inline src/libs/actions/IOU.ts:2191]: Let's avoid empty object default values, is this necessary?

File: src/pages/iou/request/IOURequestStartPage.tsx [hunk 4]
```diff
@@ -83,11 +71,11 @@ function IOURequestStartPage({
     const transactionRequestType = useRef(TransactionUtils.getRequestType(transaction));
     const previousIOURequestType = usePrevious(transactionRequestType.current);
     const {canUseP2PDistanceRequests} = usePermissions();
-    const isFromGlobalCreate = _.isEmpty(report.reportID);
+    const isFromGlobalCreate = _.isEmpty(report?.reportID);
```
Comment (blazejkustra) [inline src/pages/iou/request/IOURequestStartPage.tsx:74]: Remove underscore usage, instead use isEmptyObject utility

File: src/pages/iou/request/IOURequestStartPage.tsx [hunk 5]
```diff
@@ -102,38 +90,38 @@ function IOURequestStartPage({
...
     }, [transaction, policy, reportID, iouType, isFromGlobalCreate]);
 
     const isExpenseChat = ReportUtils.isPolicyExpenseChat(report);
     const isExpenseReport = ReportUtils.isExpenseReport(report);
-    const shouldDisplayDistanceRequest = iouType !== CONST.IOU.TYPE.TRACK_EXPENSE && (canUseP2PDistanceRequests || isExpenseChat || isExpenseReport || isFromGlobalCreate);
+    const shouldDisplayDistanceRequest = iouType !== CONST.IOU.TYPE.TRACK_EXPENSE && (!!canUseP2PDistanceRequests || isExpenseChat || isExpenseReport || isFromGlobalCreate);
 
     // Allow the user to create the request if we are creating the request in global menu or the report can create the request
-    const isAllowedToCreateRequest = _.isEmpty(report.reportID) || ReportUtils.canCreateRequest(report, policy, iouType);
+    const isAllowedToCreateRequest = _.isEmpty(report?.reportID) || ReportUtils.canCreateRequest(report, policy, iouType);
```
Comment (blazejkustra) [inline src/pages/iou/request/IOURequestStartPage.tsx:104]: Same here

File: src/pages/iou/request/step/IOURequestStepAmount.tsx (hunk 1 above)
Comment (blazejkustra) [reply src/pages/iou/request/step/IOURequestStepAmount.tsx:62]: Let's not refactor it here, unless you think it's a 10 mins fix

File: src/pages/iou/request/step/IOURequestStepAmount.tsx [hunk 6]
```diff
@@ -118,10 +108,14 @@ function IOURequestStepAmount({
...
     };
 
+    type AmountParams = {
+        amount: string;
+    };
+
     /**
-     * @param {Number} amount
+     * @param amount
      */
```
Comment (blazejkustra) [inline src/pages/iou/request/step/IOURequestStepAmount.tsx:117]: You can remove it completely

File: src/pages/iou/request/step/IOURequestStepAmount.tsx [hunk 7]
```diff
@@ -118,10 +108,14 @@ function IOURequestStepAmount({
         Navigation.navigate(ROUTES.MONEY_REQUEST_STEP_CURRENCY.getRoute(action, iouType, transactionID, reportID, backTo ? 'confirm' : '', Navigation.getActiveRouteWithoutParams()));
     };
 
+    type AmountParams = {
+        amount: string;
+    };
```
Comment (blazejkustra) [inline src/pages/iou/request/step/IOURequestStepAmount.tsx:113]: Let's move it above the component

File: src/pages/iou/request/step/IOURequestStepAmount.tsx [hunk 8]
```diff
@@ -175,11 +169,11 @@ function IOURequestStepAmount({
             headerTitle={translate('iou.amount')}
             onBackButtonPress={navigateBack}
             testID={IOURequestStepAmount.displayName}
-            shouldShowWrapper={Boolean(backTo || isEditing)}
+            shouldShowWrapper={Boolean(backTo) || isEditing}
```
Comment (blazejkustra) [inline src/pages/iou/request/step/IOURequestStepAmount.tsx:172]:
```suggestion
            shouldShowWrapper={!!backTo || isEditing}
```

File: src/pages/iou/request/step/IOURequestStepAmount.tsx [hunk 9]
```diff
@@ -175,11 +169,11 @@ function IOURequestStepAmount({
...
             includeSafeAreaPaddingBottom
         >
             <MoneyRequestAmountForm
-                isEditing={Boolean(backTo || isEditing)}
+                isEditing={Boolean(backTo) || isEditing}
```
Comment (blazejkustra) [inline src/pages/iou/request/step/IOURequestStepAmount.tsx:176]:
```suggestion
                isEditing={!!backTo || isEditing}
```

File: src/pages/iou/request/step/IOURequestStepAmount.tsx [hunk 10]
```diff
@@ -191,25 +185,23 @@ function IOURequestStepAmount({
     );
 }
 
-IOURequestStepAmount.propTypes = propTypes;
-IOURequestStepAmount.defaultProps = defaultProps;
 IOURequestStepAmount.displayName = 'IOURequestStepAmount';
 
 export default compose(
```
Comment (blazejkustra) [inline src/pages/iou/request/step/IOURequestStepAmount.tsx:190]: We can't use compose in TS files, check how other IOU pages do it

**Review by blazejkustra — CHANGES_REQUESTED** [review]

File: src/pages/iou/request/step/IOURequestStepAmount.tsx (hunk 1 above)
Comment (kosmydel) [reply src/pages/iou/request/step/IOURequestStepAmount.tsx:62]: I noticed this approach being used in multiple places, so I will keep it as is. However, we should consider using the `useAutoFocusInput` hook throughout the entire app to clean things up.

File: src/libs/Navigation/types.ts (hunk 2 above)
Comment (kosmydel) [reply src/libs/Navigation/types.ts:437]: Hm, running `npm run typecheck` results in errors for me after removing it:
```
src/pages/iou/request/step/IOURequestStepAmount.tsx:47:52 - error TS2339: Property 'backTo' does not exist on type 'Readonly<{ iouType: ValueOf<{ readonly SEND: "send"; readonly SPLIT: "split"; readonly REQUEST: "request"; readonly TRACK_EXPENSE: "track-expense"; }>; reportID: string; transactionID: string; } | { ...; }>'.

47         params: {iouType, reportID, transactionID, backTo, action},
                                                      ~~~~~~

src/pages/iou/request/step/IOURequestStepAmount.tsx:47:60 - error TS2339: Property 'action' does not exist on type 'Readonly<{ iouType: ValueOf<{ readonly SEND: "send"; readonly SPLIT: "split"; readonly REQUEST: "request"; readonly TRACK_EXPENSE: "track-expense"; }>; reportID: string; transactionID: string; } | { ...; }>'.

47         params: {iouType, reportID, transactionID, backTo, action},
                                                              ~~~~~~

src/pages/iou/request/step/IOURequestStepDistance.tsx:51:18 - error TS2339: Property 'action' does not exist on type 'Readonly<{ iouType: ValueOf<{ readonly SEND: "send"; readonly SPLIT: "split"; readonly REQUEST: "request"; readonly TRACK_EXPENSE: "track-expense"; }>; reportID: string; transactionID: string; } | { ...; }>'.

51         params: {action, iouType, reportID, transactionID, backTo},
                    ~~~~~~

src/pages/iou/request/step/IOURequestStepDistance.tsx:51:60 - error TS2339: Property 'backTo' does not exist on type 'Readonly<{ iouType: ValueOf<{ readonly SEND: "send"; readonly SPLIT: "split"; readonly REQUEST: "request"; readonly TRACK_EXPENSE: "track-expense"; }>; reportID: string; transactionID: string; } | { ...; }>'.

51         params: {action, iouType, reportID, transactionID, backTo},
                                                              ~~~~~~


Found 4 errors in 2 files.

Errors  Files
     2  src/pages/iou/request/step/IOURequestStepAmount.tsx:47
     2  src/pages/iou/request/step/IOURequestStepDistance.tsx:51
```

**Review by blazejkustra — APPROVED** [review]
Looks good!

File: src/libs/Navigation/types.ts (hunk 2 above)
Comment (blazejkustra) [reply src/libs/Navigation/types.ts:437]: Okay, I have the same now. I think it's fine to leave `never` types here

File: src/components/DistanceRequest/DistanceRequestRenderItem.tsx [hunk 11]
```diff
@@ -17,7 +17,7 @@ type DistanceRequestProps = {
     onSecondaryInteraction?: () => void;
 
     /** Function to get the index of the item */
-    getIndex?: () => number;
+    getIndex?: () => number | undefined;
```
Comment (kubabutkiewicz) [inline src/components/DistanceRequest/DistanceRequestRenderItem.tsx:20]: why do we need that extra `undefined` in type?

File: src/pages/iou/request/IOURequestStartPage.tsx [hunk 12]
```diff
@@ -161,7 +149,6 @@ function IOURequestStartPage({
                             {iouType !== CONST.IOU.TYPE.SEND ? (
                                 <OnyxTabNavigator
                                     id={CONST.TAB.IOU_REQUEST_TYPE}
-                                    selectedTab={selectedTab || CONST.IOU.REQUEST_TYPE.SCAN}
```
Comment (kubabutkiewicz) [inline src/pages/iou/request/IOURequestStartPage.tsx:164]: why this is removed ?

File: src/pages/iou/request/step/IOURequestStepAmount.tsx [hunk 13]
```diff
@@ -56,22 +49,22 @@ function IOURequestStepAmount({
...
     const isSaveButtonPressed = useRef(false);
-    const originalCurrency = useRef(null);
+    const originalCurrency = useRef<string | null>(null);
     const iouRequestType = getRequestType(transaction);
     const isEditing = action === CONST.IOU.ACTION.EDIT;
     const isSplitBill = iouType === CONST.IOU.TYPE.SPLIT;
     const isEditingSplitBill = isEditing && isSplitBill;
-    const {amount: transactionAmount} = ReportUtils.getTransactionDetails(isEditingSplitBill && !lodashIsEmpty(splitDraftTransaction) ? splitDraftTransaction : transaction);
-    const {currency} = ReportUtils.getTransactionDetails(isEditing ? draftTransaction : transaction);
+    const {amount: transactionAmount} = ReportUtils.getTransactionDetails(isEditingSplitBill && !lodashIsEmpty(splitDraftTransaction) ? splitDraftTransaction : transaction) ?? {amount: 0};
```
Comment (kubabutkiewicz) [inline src/pages/iou/request/step/IOURequestStepAmount.tsx:62]: Why we can't use `isEmptyObject`?

File: src/components/DistanceRequest/DistanceRequestRenderItem.tsx (hunk 11 above)
Comment (kosmydel) [reply src/components/DistanceRequest/DistanceRequestRenderItem.tsx:20]: This type is coming from the `DraggableList` (`RenderItemParams<T>`):
```typescript
export declare type RenderItemParams<T> = {
    item: T;
    getIndex: () => number | undefined;
    drag: () => void;
    isActive: boolean;
};
```

File: src/pages/iou/request/IOURequestStartPage.tsx (hunk 12 above)
Comment (kosmydel) [reply src/pages/iou/request/IOURequestStartPage.tsx:164]: The `OnyxTabNavigator` component doesn't accept this prop. It is loaded using `withOnyx` inside the component.

File: src/pages/iou/request/step/IOURequestStepAmount.tsx (hunk 13 above)
Comment (kosmydel) [reply src/pages/iou/request/step/IOURequestStepAmount.tsx]: Replaced :D

**Review by kubabutkiewicz — APPROVED** [review]

**Comment by kosmydel** [comment]
> Bug: Dragging and droping waypoints seems broken

I've fixed it. It is ready for review :)

---

However, when testing, I encountered a bug:
1. Log in to the account.
2. FAB -> Request Money -> Distance
3. Select a waypoint
4. Exit the RHP
5. Again point 2.
6. Press on the `Start` waypoint
7. Not found view.

Video (`main` branch): 

https://github.com/Expensify/App/assets/104823336/9830e07c-1917-4556-ab70-ac60af3ed885

**Review by ishpaul777 — APPROVED** [review]
LGTM!

**Review by luacmartins — APPROVED** [review]

**Comment by blazejkustra** [comment]
@ishpaul777 Reviewer checklist is failing

**Comment by brunovjk** [comment]
This PR removed the `selectedTab` prop from `OnyxTabNavigator`, as it already pulls it from Onyx, [comment](https://github.com/Expensify/App/pull/39910/files#r1557965387). However, we still needed to retain a default value set to `SCAN` for cases where Onyx doesn't have a `selectedTab` value. The change inadvertently caused an issue where [the default selected tab should be set to "Scan" for submitting expenses](https://github.com/Expensify/App/issues/46916).

**Pull request merged**